### PR TITLE
D8CORE-1241: added tests for authenticated user permissions

### PR DESCRIPTION
**Pull request by imonroe**
# READY FOR REVIEW

# Summary
- D8CORE-1241 calls for automated tests for security vulnerabilities for authenticated users.
- Codeception was specified as the testing platform.
- Here we have a basic set of tests for authenticated users to make sure they can't do admin things, and to make sure they do have permissions on things which they have been granted. 

# Needed By (Date)
- next codefreeze.

# Urgency
- not critical.

# Steps to Test

1. Who tests the tests, man?
image

# Affected Projects or Products
- If it's good for stanford_profile, I'll move it to soe_profile as well.

### CODE REVIEW

**Comment by sherakama** [comment]
@jbickar I am looking to you to validate that this meets the scope of the ticket.

**Comment by jbickar** [comment]
Missed this when it came through. I will take a look.

**Comment by jbickar** [comment]
Per [this comment in D8CORE-1241](https://stanfordits.atlassian.net/browse/D8CORE-1241?focusedCommentId=89515&oldIssueView=true&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-89515), these are the use cases to test for, if possible:

- Site Manager cannot escalate own role above Site Manager
- Site Manager cannot escalate the role of others above Site Manager
- PHP entered in a redirect gets sanitized
- PHP code gets sanitized on content creation
- Try to upload a PHP file in Media and it fails form validation
- Try to upload a PHP file as a favicon in theme settings and have it fail (error message for me was "For security reasons, your upload has been renamed to foo.php.txt.")
- Try to upload a PHP file as a logo in theme settings and have it fail (error message for me was "The image file is invalid or the image type is not allowed. Allowed types: gif, jpe, jpeg, jpg, png")

**Comment by imonroe** [comment]
Aha, sorry about that; authenticated users is a particular role, without many permissions at all above the anonymous user role, so I was testing for that.  I'll add additional tests per above.

**Comment by imonroe** [comment]
Ok, I successfully added additional tests for:
- Site Manager cannot escalate their own role above Site Manager.
- Site Manager cannot escalate others' role above Site Manager.
- PHP code is not allowed in redirects.
- PHP code is escaped and not run when added to content.
- PHP files are not allowed as uploads for favicons.
- PHP files are not allowed as uploads for the logo.

I was unable to find a way to test the uploading of PHP files in the media browser, since that's a React JS component.  There might be a way to do so using functional tests, but I was unable to make it work.

File: tests/codeception/acceptance/AuthenticatedPermissionsCest.php [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+/**
+ * Test the restrictions on authenticated users.
+ */
+class AuthenticatedPermissionsCest {
+
+  /**
+   * Set up a file to test PHP injection.
+   */
+  public function _before(AcceptanceTester $I) {
+    file_put_contents(codecept_data_dir() . '/injection.php', '<?php echo("injection test"); die(); ?>');
```
Comment (sherakama) [inline tests/codeception/acceptance/AuthenticatedPermissionsCest.php:12]: OOOOOH... I like this.

I am going to steal this for another issue I have been having with the lockup logo. @pookmish

Comment (sherakama) [reply tests/codeception/acceptance/AuthenticatedPermissionsCest.php:12]: See: https://github.com/SU-SWS/stanford_profile/pull/220/files#diff-57971af0c9829dc91faff7b65b36951cR300-R308

File: tests/codeception/acceptance/AuthenticatedPermissionsCest.php [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  }
+
+  /**
+   * Remove the php injection file.
+   */
+  public function _after(AcceptanceTester $I) {
+    unlink(codecept_data_dir() . '/injection.php');
```
Comment (sherakama) [inline tests/codeception/acceptance/AuthenticatedPermissionsCest.php:19]: This is SMRT.

File: tests/codeception/acceptance/AuthenticatedPermissionsCest.php (hunk 1 above)
Comment (pookmish) [reply tests/codeception/acceptance/AuthenticatedPermissionsCest.php:12]: huh that could be a nice way to get images for usage.

File: tests/codeception/acceptance/AuthenticatedPermissionsCest.php (hunk 2 above)
Comment (pookmish) [reply tests/codeception/acceptance/AuthenticatedPermissionsCest.php:19]: I cant remember, does `_after` run on both failures and successes?

**Comment by sherakama** [comment]
LG2M 👍

File: tests/codeception/acceptance/AuthenticatedPermissionsCest.php (hunk 2 above)
Comment (imonroe) [reply tests/codeception/acceptance/AuthenticatedPermissionsCest.php:19]: The docs say both `_before` and `_after` get run for every test; it doesn't explicitly say the outcome of the tests matters.  I believe it runs no matter if it's a success or a failure.

**Comment by pookmish** [comment]
No need for that ticket. A test already exists https://github.com/SU-SWS/stanford_profile/blob/8.x-1.x/tests/codeception/functional/Paragraphs/WYSIWYGCest.php#L143

**Comment by jbickar** [comment]
Ah, nice.

**Comment by imonroe** [comment]
well, there are two approvals, the additional test already exists, so I'm going to go ahead and merge this.

**Pull request merged**